<img src="http://escience.washington.edu/wp-content/uploads/2016/02/DataScienceForSocialGood.png" width = "550">


### SQL Tutorial 

#### Friday June 17, 9 am - 12 pm

#### Instructors
Valentina Staneva, Bill Howe, Anthony Arendt



#### 1) How many "TRESPASS" offenses occurred in total?
<br>
#### 2) What is the most common crime in the dataset, and in what year and month did these crimes occur?
<br>
#### 3) What is the most common crime within 5 km of my house?
<br><br>
<small>Data source: "SeattleCrimeIncidents.csv" or http://www.seattle.gov/seattle-police-department/crime-data/spd-data-sets

#### What is a database? 
* software system for capturing, storing and analyzing data 
* nearly all databases use the _relational_ data model

<img src='images/terminology.png' width = 900px>

#### Relational data model:
* data are structured into row/column format 

| crimesID | Offense type | Offense code | Date | Location | 
| ---- | ---- | ----- | ---- | ---- | ---- |
|  <small> 1 | <small>tresspass | <small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 2 | <small>larceny-theft | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St | 

* each record has a unique identifier (primary key)

#### Relational data model:
* uses Structured Query Language (SQL):   

```SQL
CREATE TABLE seattlecrimesincidents 
    ("crimesID" int,
     "Offense type" character
     "Offense code" int
     "Date" timestamp
     "Location" character); 
```

* populating the database records:
```SQL
INSERT INTO seattlecrimeincidents VALUES

    (1,'trespass', 5700,'2015-01-28 09:30:00','12XX Block of E Pike St'),
    
    (2,'larceny-theft',2300, '2015-02-21 08:24:21','15XX Block of Aurora St');
```

| crimesID | Offense type | Offense code | Date | Location | 
| ---- | ---- | ----- | ---- | ---- | ---- |
|  <small> 1 | <small>tresspass | <small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 2 | <small>larceny-theft | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St | 

### Data in each column must be of the same type

#### SQL requires this so it knows how to operate on the data

Some common [data types](https://www.postgresql.org/docs/9.4/static/datatype.html):
<br>
<div class="TABLE">
    <a name="DATATYPE-TABLE" id="DATATYPE-TABLE"></a>

<small>
    <table class="CALSTABLE" border="1">
      <colgroup><col>
      <col>
      <col>

      <thead>
        <tr>
          <th>Name</th>

          <th>Aliases</th>

          <th>Description</th>
        </tr>
      </thead>

      <tbody>

        <tr>
          <td><tt class="TYPE">boolean</tt></td>

          <td><tt class="TYPE">bool</tt></td>

          <td>logical Boolean (true/false)</td>
        </tr>

        <tr>
          <td><tt class="TYPE">character [(<tt class="REPLACEABLE c4">n</tt>)]</tt></td>

          <td><tt class="TYPE">char [(<tt class="REPLACEABLE c4">n</tt>)]</tt></td>

          <td>fixed-length character string</td>
        </tr>

        <tr>
          <td><tt class="TYPE">date</tt></td>

          <td>&nbsp;</td>

          <td>calendar date (year, month, day)</td>
        </tr>

        <tr>
          <td><tt class="TYPE">double precision</tt></td>

          <td><tt class="TYPE">float8</tt></td>

          <td>double precision floating-point number (8 bytes)</td>
        </tr>

        <tr>
          <td><tt class="TYPE">integer</tt></td>

          <td><tt class="TYPE">int</tt>, <tt class="TYPE">int4</tt></td>

          <td>signed four-byte integer</td>
        </tr>

        <tr>
          <td><tt class="TYPE">json</tt></td>

          <td>&nbsp;</td>

          <td>JSON data</td>
        </tr>


        <tr>
          <td><tt class="TYPE">money</tt></td>

          <td>&nbsp;</td>

          <td>currency amount</td>
        </tr>

        <tr>
          <td><tt class="TYPE">timestamp [(<tt class="REPLACEABLE c4">p</tt>)] [ without time zone
          ]</tt></td>

          <td>&nbsp;</td>

          <td>date and time (no time zone)</td>
        </tr>

        <tr>
          <td><tt class="TYPE">xml</tt></td>

          <td>&nbsp;</td>

          <td>XML data</td>
        </tr>
      </tbody>
    </table>
  </div>

* selecting data:
```SQL
SELECT * 
   FROM seattlecrimeincidents 
   WHERE "Offense code" = 5700;
```

| crimesID | Offense type | Offense code | Date | Location | 
| ---- | ---- | ----- | ---- | ---- | ---- |
|  <small> 1 | <small>tresspass | <small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |

### Database rules:

#### take the time to [normalize](https://en.wikipedia.org/wiki/Database_normalization) your tables to minimize redundancy

#### example: multiple offenses at the same time 

| crimesID | Offense type | Offense code | Date | Location | 
| ---- | ---- | ----- | ---- | ---- | ---- |
|  <small> 1 | <small>tresspass or burglary | <small>5700 or 5710 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 2 | <small>larceny-theft | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St | 

####  INCORRECT: database will have problems searching these columns

#### solution: create another row

| crimesID | Offense type | Offense code | Date | Location | 
| ---- | ---- | ----- | ---- | ---- | ---- |
|  <small> 1 | <small>tresspass | <small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 2 | <small> burglary | <small> 5710 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 3 | <small>larceny-theft | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St |

### NULL values

* missing data are a common feature of many datasets
* here the code for "tresspass" is not known so the data entry is "X"

| crimesID | Offense type | Offense code | Date | Location | 
| ---- | ---- | ----- | ---- | ---- | ---- |
|  <small> 1 | <small>tresspass | <small>X | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 2 | <small> burglary | <small> 5710 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 3 | <small>larceny-theft | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St |




#### NULL values
* conventionally, some value is used to represent missing data (e.g. "X" or -9999) 
* relational databases introduced NULL values:
    * NULL is a state representing a lack of a value
    * NULL is not the same as zero!
    * NULL values are ignored in SELECT statements

#### Joining Tables
* well designed databases distribute data across multiple tables, for efficiency
* then we can JOIN data between tables as needed

| crimesID |  Offense code | Date | Location | 
| ---- |  ----- | ---- | ---- | ---- |
|  <small> 1 |<small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 1 |<small>5700 | <small> 2015-02-12 03:25:00 | <small> 1XX Block of Aloha St |
|  <small> 2 | <small> 5710 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |
|  <small> 3 | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St |

<br>

| typesID | Offense type | Offense code | 
| ---- | ---- | ----- | ---- | |
|  <small> 1 | <small>tresspass | <small>5700 | 
|  <small> 2 | <small> burglary | <small> 5710 | 
|  <small> 3 | <small>larceny-theft | <small>2300 | 

<img src = 'images/joinTables.png' width = 800px>

#### Data Analysis:
* databases have powerful methods for analyzing data
* one of the most common tasks: applying statistics across groups
* SQL syntax: GROUP BY

| crimesID |  Offense code | Date | Location | Damage | 
| ---- |  ----- | ---- | ---- | ---- | ---- |
|  <small> 1 |<small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |<small> \$1,220 | 
|  <small> 1 |<small>5700 | <small> 2015-02-12 03:25:00 | <small> 1XX Block of Aloha St |<small> \$11,420 |
|  <small> 2 | <small> 5710 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |<small> \$5,389 |
|  <small> 2 | <small> 5710 | <small> 2015-1-02 12:31:20 | <small> 12XX Block of E Pine St |<small> \$15,231 |
|  <small> 3 | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St |<small> \$2,405 |

* data grouped by "Offense code":

| crimesID |  Offense code | Date | Location | Damage | 
| ---- |  ----- | ---- | ---- | ---- | ---- |
|  <small> 1 |<small>5700 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |<small> \$1,220 | 
|  <small> 2 |<small>5700 | <small> 2015-02-12 03:25:00 | <small> 1XX Block of Aloha St |<small> \$11,420 |
|    |    |    |    |    |
|  <small> 3 | <small> 5710 | <small> 2015-01-28 09:30:00 | <small> 12XX Block of E Pike St |<small> \$5,389 |
|  <small> 4 | <small> 5710 | <small> 2015-1-02 12:31:20 | <small> 12XX Block of E Pine St |<small> \$15,231 |
|    |    |    |    |    |
|  <small> 5 | <small>2300 | <small> 2015-02-21 08:24:21 | <small> 15XX Block of Aurora St |<small> \$2,405 |

* result of SUM(Damage) over "Offense code" groups:

| Offense code | totalDamage | 
|   ---- | ---- |
|  <small>5700  |<small> \$12,640 | 
|  <small> 5710 |<small> \$20,620 |
|  <small>2300  |<small> \$2,405 |

#### Database Implementation:

* there are many relational database software implementations:
   * commercial: Oracle, Microsoft SQL Server, IBM DB2 
   * open source: MySQL, PostgreSQL

* regardless of the software:
   * most databases are deployed on a server 
   * can deploy locally for testing


#### Database Interface:

* all databases are accessed via a _connection string_:
   * hostname, port, user, password


<img src='images/databaseDiagram.png' width = 800px>

# EXERCISE 1: 15-20 MINUTES

### EXERCISE 1 DEBRIEF:

* what did you learn? 
* what was challenging or surprised you?

### Recall our second analysis question:

* "What is the most common crime in the dataset, and in what year and month did these crimes occur?"



#### To solve this we need to learn more about GROUP BY and the "date" field  

In [2]:
%load_ext sql
%sql postgresql://awsdssg:datascience2016@dssg2016.cel0foggtnsk.us-east-1.rds.amazonaws.com/dssg2016

'Connected: awsdssg@dssg2016'

In [8]:
%%sql
SELECT "Offense Type", count(*) AS incident_count 
FROM seattlecrimeincidents
GROUP BY "Offense Type" 
ORDER BY "incident_count" DESC
limit 10;

10 rows affected.


[('THEFT-CARPROWL', 5334),
 ('THEFT-OTH', 1828),
 ('VEH-THEFT-AUTO', 1684),
 ('PROPERTY DAMAGE-NON RESIDENTIA', 1412),
 ('BURGLARY-FORCE-RES', 1361),
 ('DISTURBANCE-OTH', 1356),
 ('ASSLT-NONAGG', 1210),
 ('THEFT-SHOPLIFT', 1209),
 ('PROPERTY FOUND', 1063),
 ('FRAUD-IDENTITY THEFT', 856)]

In [9]:
%%sql
SELECT "Offense Type", "Year", "Month", count(*) AS incident_count 
FROM seattlecrimeincidents 
GROUP BY "Offense Type", "Year", "Month" 
ORDER BY "incident_count" DESC
limit 10;

10 rows affected.


[('THEFT-CARPROWL', 2015, 1, 1163),
 ('THEFT-CARPROWL', 2015, 5, 949),
 ('THEFT-CARPROWL', 2015, 4, 829),
 ('THEFT-CARPROWL', 2015, 3, 800),
 ('THEFT-CARPROWL', 2015, 2, 783),
 ('THEFT-CARPROWL', 2015, 6, 734),
 ('VEH-THEFT-AUTO', 2015, 1, 422),
 ('THEFT-OTH', 2015, 1, 390),
 ('BURGLARY-FORCE-RES', 2015, 1, 316),
 ('THEFT-OTH', 2015, 3, 300)]

## EXERCISE 2: 20-30 MINUTES

#### Exercise 2 Debrief:

* what did you learn?

### Recall our final analysis question:

* "What is the most common crime within 5 km of my house?"

#### To answer this we'll need to understand something about mapping, and how databases encode spatial information.

### Note that the database currently has latitude and longitude information:

In [10]:
%sql SELECT "Latitude", "Longitude" FROM seattlecrimeincidents WHERE gid < 5;

4 rows affected.


[(47.6158384, -122.3181689),
 (47.60087709, -122.3312162),
 (47.59582098, -122.3175691),
 (47.6140991, -122.3174884)]

#### Question: what is the straight line distance between points 1 and 2?

| Point | Latitude | Longitude |
| ---- | ---- | ---- |
| 1 | 47.62 | -122.32 |
| 2 | 47.58 | -122.33 | 


<br><br>
<img src="images/earthLatLong.png" width = "800">
<br><br><br><br>
[source](https://coast.noaa.gov/digitalcoast/training/datums)

<img src="http://desktop.arcgis.com/en/arcmap/10.3/guide-books/map-projections/GUID-C7DB60DD-58FF-4D48-8FBE-85224C8FABD1-web.gif" width = 500>

### How does the database currently encode the latitude and longitude information?

In [19]:
%%sql
SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'seattlecrimeincidents' AND (column_name = 'Latitude' OR column_name = 'Longitude');

2 rows affected.


[('Longitude', 'double precision'), ('Latitude', 'double precision')]

### It would be better if the database understands latitude and longitude as locations rather than as double precision numbers.

#### Fortunately databases have a special data type called _geometries_ that encode the location of spatial (_vector_) information. 

<img src="http://neon-workwithdata.github.io/NEON-R-Spatio-Temporal-Data-and-Management-Intro/images/dc-spatial-vector/pnt_line_poly.png" width = "500">

http://neon-workwithdata.github.io/NEON-R-Spatio-Temporal-Data-and-Management-Intro/